adam_preprocessing/shape_stroke_extraction.py

```diff
@@ -567,7 +567,18 @@ def stroke_extraction_from_matlab(self):
         eng = matlab.engine.start_matlab()
         file_dir = os.path.dirname(__file__)
         eng.addpath(eng.genpath(file_dir), nargout=0)
-        out = eng.ske(self.path, nargout=2)
+        if self.process_masks_independently:
+            s_out, e_out = [], []
+            for segmentation_filename in self.distinct_mask_filenames:
+                path = os.path.join(self.output_dir, segmentation_filename)
+                s: List[matlab.double]
+                e: matlab.double
+                s, e = eng.ske(path, nargout=2)
+                s_out.extend(s)
+                e_out.append(e)
+            out = (s_out, eng.cat(1, *e_out))
+        else:
+            out = eng.ske(self.path, nargout=2)
```

Comment:
@spigo900 is this roughly what I should be doing here? Extracting all the strokes separately from each masks and then putting them all together into one big array?

Also, it's weird to me that one of the outputs of `ske()` is a `list[matlab.double]` while the other output is just a `matlab.double` (I probably just don't know enough about MATLAB/its Python API to see why this makes sense)

Reply:
No, this won't do what we want. The stroke graphs will start separate but get glued together during the post-processing. The post-processing step adds edges to the stroke graph between strokes that are sufficiently close to each other.

What we want to do is to run, for each image, the whole pipeline up until the part where we convert to YAML. This produces some outputs that get fed into the "convert to YAML" step. I think we want to concatenate those outputs and then feed them to "convert to YAML." I think this should be relatively straightforward though a bit annoying to do. I'm thinking you'll want to actually concatenate them since that keeps the code the same. Some thoughts on concatenating the five implicit inputs to "convert to YAML," that is `num_obj`, `stroke_obj_ids`, `reduced_stroke`, `adj`, and `color`:

1. `num_obj` you can just sum, of course.
2. `stroke_obj_ids` is a NumPy array and needs special treatment. Before concatenating `prev_stroke_obj_ids` and `new_stroke_obj_ids` you'll want to increment `new_stroke_obj_ids` by `prev_num_objs` (aka `np.max(prev_stroke_obj_ids) + 1`?) to avoid accidentally gluing together different objects. I expect stroke object IDs to start at 0 and increase to `num_objs - 1`, with each mask's stroke object IDs starting at 0 by default such that you have to manually increment them before concatenating.
3. `reduced_strokes` you can just straight up concatenate, I think. It's a NumPy array.
4. You'll probably want to concatenate the adjacency matrices along the diagonal. That way you don't have to change the existing logic. I am not sure the easiest/most efficient way to do this in NumPy. Worst case you can allocate a new adjacency matrix using `cat_adj = np.zeros()` and use `cat_adj[:self.num_objs, :][:, :self.num_objs] = old_adj; cat_adj[:self.num_objs, :][:, :self.num_objs] = new_adj`. Lots of allocations and uses lots of memory but it should work OK for us.
5. `colors` is 2D of shape `[num_obj, 5]`. I think you can just concatenate on the first axis.

I don't understand the MATLAB API that well myself, so not sure why it's a list of MATLAB doubles.